### PR TITLE
Drop stuck close execution transfer task

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
Drop stuck close execution transfer task due to missing workflow start event

**Why?**
Lots of corrupted workflow history.  

**How did you test it?**
Unit test.

**Potential risks**
N/A

### CODE REVIEW

File: common/service/dynamicconfig/config.go [hunk 1]
```diff
@@ -289,6 +292,18 @@ func (c *Collection) GetBoolPropertyFnWithDomainFilter(key Key, defaultValue boo
 	}
 }
 
+// GetBoolPropertyFnWithDomainIDFilter gets property with domainID filter and asserts that it's a bool
+func (c *Collection) GetBoolPropertyFnWithDomainIDFilter(key Key, defaultValue bool) BoolPropertyFnWithDomainIDFilter {
```
Comment (yycptt) [inline common/service/dynamicconfig/config.go:296]: Will create similar functions for other data types if decided to merge into master.

File: common/metrics/defs.go [hunk 2]
```diff
@@ -1965,6 +1967,7 @@ var MetricDefs = map[ServiceIdx]map[int]metricDefinition{
 		TaskLimitExceededCounter:                          {metricName: "task_errors_limit_exceeded_counter", metricType: Counter},
 		TaskProcessingLatency:                             {metricName: "task_latency_processing", metricType: Timer},
 		TaskQueueLatency:                                  {metricName: "task_latency_queue", metricType: Timer},
+		StuckTaskDiscardedCounter:                         {metricName: "stuck_task_discarded_counter", metricType: Counter},
```
Comment (vancexu) [inline common/metrics/defs.go:1970]: I would be specific for this like "transfer_task_missing_event_counter"

File: service/history/service.go [hunk 3]
```diff
@@ -205,6 +205,8 @@ type Config struct {
 
 	//Crocess DC Replication configuration
 	ReplicationEventsFromCurrentCluster dynamicconfig.BoolPropertyFnWithDomainFilter
+
+	EnableDropStuckTaskByDomain dynamicconfig.BoolPropertyFnWithDomainIDFilter
```
Comment (vancexu) [inline service/history/service.go:209]: EnableDropStuckTaskByDomainID

File: service/history/taskProcessor.go [hunk 4]
```diff
@@ -270,6 +270,15 @@ func (t *taskProcessor) handleTaskError(
 		return nil
 	}
 
+	if transferTask, ok := task.task.(*persistence.TransferTaskInfo); ok &&
+		transferTask.TaskType == persistence.TransferTaskTypeCloseExecution &&
+		err == ErrMissingWorkflowStartEvent &&
+		t.config.EnableDropStuckTaskByDomain(task.task.GetDomainID()) { // use domainID here to avoid accessing domainCache
+		scope.IncCounter(metrics.StuckTaskDiscardedCounter)
+		task.logger.Error("Drop close execution transfer task due to corrupted workflow history", tag.Error(err), tag.LifeCycleProcessingFailed)
```
Comment (vancexu) [inline service/history/taskProcessor.go:278]: let's add log of workflow info for auditing

Comment (yycptt) [reply service/history/taskProcessor.go:278]: Task logger is initialized with tags for domainID, workflowID, runID, etc.

File: common/metrics/defs.go (hunk 2 above)
Comment (yycptt) [reply common/metrics/defs.go]: Done.

File: service/history/service.go (hunk 3 above)
Comment (yycptt) [reply service/history/service.go]: Done.

**Pull request merged**